java/java-impl/src/com/intellij/codeInsight/FunctionalInterfaceSuggester.java

```diff
@@ -182,77 +184,149 @@
+    final PsiParameter[] parameters = interfaceMethod.getParameterList().getParameters();
+    final PsiType[] left = new PsiType[parameters.length + 1];
+    final PsiType[] right = new PsiType[parameters.length + 1];
+    final PsiMethod method = getTargetMethod(methodReferenceExpression, qualifierType, parameters, left, right);
+    if (method == null) {
+      return null;
+    }
 
-        final PsiSubstitutor substitutor = PsiResolveHelper.SERVICE.getInstance(interface2Consider.getProject())
-          .inferTypeArguments(interface2Consider.getTypeParameters(), left, right, PsiUtil.getLanguageLevel(expression));
+    PsiParameter[] functionalExprParameters = method.getParameterList().getParameters();
 
-        PsiType type = elementFactory.createType(interface2Consider, substitutor);
+    left[parameters.length] = method.getReturnType();
+    right[parameters.length] = interfaceMethod.getReturnType();
 
-        if (expression.isAcceptable(type)) {
-          return type;
-        }
+    final int offset;
+    if (PsiMethodReferenceUtil.isStaticallyReferenced(methodReferenceExpression) &&
+        !method.hasModifierProperty(PsiModifier.STATIC)) {
+      offset = 1;
+    }
+    else {
+      offset = 0;
+    }
+    fillLeftRightTypes(parameters, functionalExprParameters, offset, left, right);
+
+    return substituteAcceptableType(interface2Consider, expression, elementFactory, left, right);
+  }
+
+  private static List<PsiType> composeConstructorReferenceExpressionType(PsiClass interface2Consider,
+                                                                         PsiFunctionalExpression expression,
+                                                                         PsiType qualifierType,
+                                                                         PsiElementFactory elementFactory,
+                                                                         PsiMethod interfaceMethod,
+                                                                         PsiMethodReferenceExpression methodReferenceExpression) {
+    final PsiParameter[] parameters = interfaceMethod.getParameterList().getParameters();
+    final List<ConstructorTypes> constructors = getTargetConstructors(methodReferenceExpression, parameters, qualifierType,
+                                                                      interfaceMethod.getReturnType());
+    final List<PsiType> types = new ArrayList<>();
+    for (ConstructorTypes constructorTypes : constructors) {
+      final PsiType type = substituteAcceptableType(interface2Consider, expression, elementFactory,
+                                                    constructorTypes.leftParameters, constructorTypes.rightParameters);
+      if (type != null) {
+        types.add(type);
       }
     }
-    return null;
+    return types;
+  }
+
+  @Nullable
+  private static PsiType composeLambdaExpressionType(PsiClass interface2Consider,
```

Comment:
At this context I was trying to somehow find return type of lambda-expression using com.intellij.psi.LambdaUtil#getReturnExpressions but was not able to find easy solution how to detect return type from list of PsiExpressions.